### ISSUE TITLE
ipython and python prompt

### ISSUE THREAD

**Issue by chaffra** [Alexpux/MINGW-packages#730]
The mingw version of python2 and ipython2 do not show  a prompt. I have installed the 64bit versions. Is there a way to fix this? The msys version seems to work.

**Comment by Alexpux** [comment]
Try install "winpty" ackage.

**Comment by chaffra** [comment]
I installed winpty.
Running winpty ipython from MINGW64 shell returns
"Error 0x2 starting ipython"
Maybe I should be installing ipython from pip or the problem would be 
the same?

**Comment by maoueh** [comment]
@chaffra `ipython` is probably a python script and `winpty` works only with direct executables. You can see if it's really a script by doing `file $(which ipython)`.

What I often do to workaround this is to make a bash script that supersedes the native executable and wrap it in a calls to `winpty`:

```
$ cat /home/movachon/local/bin/python
#!/usr/bin/env sh

winpty python "$@"
```

Regards,
Matt

**Comment by maoueh** [comment]
Note that this might not work and will depends on how the `ipython` script is built. For example, ruby gems have a tendancy to start with the absolute executable path like:

```
#!C:/Absolute/Path/To/Ruby/Executable/ruby.exe
```

Which breaks this tricks.

**Comment by chaffra** [comment]
Hi Matt,

Thanks for the trick, but the mingw64 version of ipython does not have a 
native executable where as the pip version does. Although the pip 
version is unable to find pyreadline on windows which is another 
problem. I guess, I am asking if it would be possible to make an 
ipython2.exe under mingw64?

**Comment by maoueh** [comment]
No, you misunderstood my trick.

Console problems is coming from the `python.exe` executable. The `ipython` script is probably using the `python.exe` to run it's own script. The trick is kinda "patch" `python.exe` to think it's running in a native console (via `winpty` which do that) and then use this to run the `ipython` script.

Let's add an example (if you can copy `ipython` script content here, it would help). I work with ruby which have the same problem.

Let's say I have a `iruby` script:

```
#!/usr/bin/env ruby

# Trimmed down version
IRB.start(__FILE__)
```

This starts a shell terminal to execute ruby code. However, without a `ruby` wrapped in a `winpty` call, this does not work correctly.

The trick is to wrap ruby with `winpty`. So I create a new `ruby` bash script which will be picked first. Then, in the script, I wrap native `ruby.exe` with `winpty`:

```
$ cat /home/movachon/local/bin/ruby
#!/usr/bin/env sh

winpty `which ruby.exe` "$@"
```

Now, when I run `irb`, the bash interpreter will use `#!/usr/bin/env ruby` to launch the script. This will resolve to my own `/home/movachon/local/bin/ruby` script (which I added to my `PATH` before original `ruby.exe` localtion). Since by bash script wraps native `ruby.exe` with `winpty`, it works correctly.

Check out [this thread](http://sourceforge.net/p/msys2/mailman/message/34285388/) on the MSYS2 mailing list about more explanation of the problem.

**Comment by chaffra** [comment]
Got it. This is my ipython script as installed by msys2
------------------------------------------------------------------------
#!/usr/bin/env python2.exe
# This script was automatically generated by setup.py
if __name__ == '__main__':
     from IPython import start_ipython
     start_ipython()
-------------------------------------------------------------------------

I created a python bash script with
--------------------------------
#!/usr/bin/env bash
winpty python "$@"
------------------------------

and I can do python /mingw64/bin/ipython

It gives me the prompt now but it's all weird characters like it's not 
loading something. The prompt is like

"←[0;32mIn [←[1;32m1←[0;32m]: ←[0m"

Any ideas?

**Comment by maoueh** [comment]
It cannot work with the current `ipython` script sadly. Problem is this line:

```
#!/usr/bin/env python2.exe
```

This means the script will always try to find executable and not your overridden bash script. 

How did you named your wrapper bash script by way? And where did you put it?

Second, your bash is not correct because you did `winpty python "$@"` which will probably simply call back your own bash script. Use `winpty $(which python2.exe) "$@"` instead.

A workaround would be to tweak in `ipython` line `#!/usr/bin/env python2.exe` to `#!/usr/bin/env python` (use the same name as your bash script).

To check that everything is correct, if you type `which python` in your terminal, it should print path to your `python` wrapper bash script.

P.S. Use triple backticks ``` to wrap your code snippet on GitHub, this will render them in boxes.

**Comment by chaffra** [comment]
Another solution is to avoid the mintty shell altogether. removing it fron msys make ipython and python prompt work. However the ipython prompt is all weird like ←[0;32mIn [←[1;32m1←[0;32m]: ←[0m

Any idea what might cause this?

**Comment by maoueh** [comment]
Yep indeed. On my workstation, I have two shells, one is `mintty` and another is a plain `cmd.exe` just for those cases. I saw that `ConEmu` is becoming a good package in `MSYS2` and could be a good replacement also but did not tried it yet.

What packages did you install exactly, I could try to reproduce here to see if I get the same output as you.

**Comment by maoueh** [comment]
Note that those are ANSI escape sequences for coloring and terminal beautifying stuff (sorry if you already knew that).

**Comment by chaffra** [comment]
Sure, I installed python2-ipython with pacman. The mingw 64bit version 
and that's it I think. Also I uninstalled mintty so that when I run 
mingw64_shell.bat it defaults to sh.exe

How can I get the escape characters to display correctly? Do you have 
the same problem?

**Comment by lazka** [comment]
Both python2 and python3 have been patched in the meantime to detect mintty as a terminal and show the prompt by default there. Not sure if this fixes the issues raised here. Testing welcome!

**Comment by MaxNoe** [comment]
Not fixed for me on python 3.6.6 using anaconda on windows

**Comment by lazka** [comment]
Is anaconda using the msys2 python?

**Comment by MaxNoe** [comment]
I don't understand the question.

Anaconda is a full python distribution.

**Comment by lazka** [comment]
Let me rephrase: What makes you think that this bug report describes your problem?

(btw, msys2 python is still at 3.6.5, so you are not using it)

**Comment by MaxNoe** [comment]
Ah sorry, I was lost in to many tabs open with issues on this topic. This is the wrong repo.

**Comment by MaxNoe** [comment]
Can you point me to *how* you fixed it? To provide the fix to anaconda / windows python?

**Comment by lazka** [comment]
https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-python3/1700-cygpty-isatty-disable-readline.patch

**Comment by MaxNoe** [comment]
This link is now broken, 
use
https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-python3/1700-cygpty-isatty.patch
